addons/mail/static/src/component_hooks/use_messaging_features/use_messaging_features.js

```diff
@@ -0,0 +1,18 @@
+/** @odoo-module */
+
+import { useModels } from "../use_models/use_models";
+import { useShouldUpdateBasedOnProps } from "../use_should_update_based_on_props/use_should_update_based_on_props";
+
+const { useComponent } = owl.hooks;
+
+export const componentRegistry = {};
```

Comment:
If there was syntax support for it I would rather have made a class decorator than a hook, this way there would be no need to expose the registry at all: components that use the decorator would get the effect of the hook and be registered in the registry at the same time. It would also avoid needing to guard against multiple calls modifying components twice. It's just so ugly in JS that I don't think we want to do it that way, but I think it would be a better technical solution.

Reply:
Instead of exporting the registry, we could export a register function.
Then registering a component could do both adding the proper hooks and the the components value, and we would have only one import/export and a single line to call on each component.

Reply:
I think what we actually want is to reach a point where we don't even need to create the component class in the code, just give a useful definition to the register/create method. That method would take care of creating the actual component class for us with the proper setup automatically.
We could also have similar method for patching components, without the need to import and patch it manually.
This is very similar to what we do for our models (model core) and the end end goal would be to completely merge the 2 together.

Reply:
> Instead of exporting the registry, we could export a register function.
> Then registering a component could do both adding the proper hooks and the the components value, and we would have only one import/export and a single line to call on each component.

So a decorator? :smile: 

> I think what we actually want is to reach a point where we don't even need to create the component class in the code, just give a useful definition to the register/create method. That method would take care of creating the actual component class for us with the proper setup automatically.

I'm not sure what you'd like to remove, the class definition already has pretty much only what's needed to create the component, the only boiler plate is the `class X extends Component` line which seems reasonable. You could absolutely pass an object containing the methods to a register method that would create the component but I'm not sure what the benefit there would be.

> We could also have similar method for patching components, without the need to import and patch it manually.

Considering templates cannot be patched dynamically, and you need to be sure that the component you're trying to patch is loaded in the registry before you try to patch it (as shown by my snailmail message debacle), I don't think there's much to gain here either.

Reply:
Don't worry about the rest for now, but if you could just make it so that registering is done with a function instead of having to manually push on the registry that would be great. And also the function should properly set up the hook automatically then.